TcUnit-Runner/VisualStudioInstance.cs

```diff
@@ -127,35 +131,61 @@ private void LoadDevelopmentToolsEnvironment(string visualStudioVersion, bool is
             tcAutomationSettings.SilentMode = true; // Only available from TC3.1.4020.0 and above
 
             // Load the correct version of TwinCAT using the remote manager in the automation interface
-            log.Info("Using the TwinCAT remote manager to load TwinCAT version '" + this.tcVersion + "'...");
             ITcRemoteManager remoteManager = dte.GetObject("TcRemoteManager");
-            
-            // If the version is pinned check if the pinned version is available
-            if (isTcVersionPinned)
+
+            string latestTcVersion = "";
+            // Check if Version is installed
+            foreach (var possibleVersion in remoteManager.Versions)
             {
-                log.Info("Project has a pinned version: " + this.tcVersion);
-                foreach (var possibleVersion in remoteManager.Versions)
+                latestTcVersion = Char.ToString(possibleVersion[0]);
+                if (possibleVersion == this.tcVersion)
                 {
-                    if (possibleVersion == this.tcVersion)
-                    {
-                        isVersionAvailable = true;
-                        break;
-                    }
+                    isVersionAvailable = true;
                 }
-                if (isVersionAvailable)
-                    log.Info("The pinned TwinCAT version is available ");
-                else
+                // Check if optional Version is installed
+                if (possibleVersion == this.forceToThisTwinCatVersion && this.forceToThisTwinCatVersion != null)
                 {
-                    log.Error("The pinned TwinCAT version is not available ");
-                    throw new Exception("The pinned TwinCAT version is not available");
+                    isForceVersionAvailable = true;
+                    log.Info("The optional TC Version is available ");
                 }
             }
 
-            if (!isTcVersionPinned || isVersionAvailable)
+            // If the Version is Pinned, check if the Pinned Version is available
+            if (isTcVersionPinned && (String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
```

Comment:
What happens if the TwinCAT version is pinned AND we have set a forced version of TwinCAT?

Reply:
... and they differ from each other?

Reply:
I don't see this being handled in the code right now

Reply:
> What happens if the TwinCAT version is pinned AND we have set a forced version of TwinCAT?

The forced version overwrites the pinned version.
So it is possible to force also pinned version to a specific version. As we discussed before
In the log you can read that the version was previously pinned to version ....
And that the build was done with the forced version ....
All the required information is available for debugging 



           

` if (isTcVersionPinned && (String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
`
> ... and they differ from each other?

> I don't see this being handled in the code right now

```
            // If the Version is Pinned, check if the Pinned Version is available
            if (isTcVersionPinned && (String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
            {

                  if (isVersionAvailable)
                  {
                      log.Info("The pinned TC Version is available ");
                  }
                  else
                  {
                      log.Error("The pinned TC Version is not available ");
                      throw new Exception("The pinned TC Version is not available");
                  }
             }
```

This code section only generates a log section if the version is pinned and the force to version option is not set.

If the version is pinned and available a log is generated.

If the version is pinned and not available the method is aborted with an exception.